relay/relay.go

```diff
@@ -263,261 +242,180 @@ func (r *Relay) process(parentCtx context.Context) error {
 		}
 	} else {
 		r.updateMetricsRelaySubDirIndex()
+		// if not a new server, try to recover the latest relay log file.
+		err = r.tryRecoverLatestFile(parser2)
+		if err != nil {
+			return errors.Trace(err)
+		}
 	}
 
-	err = r.setUpReader()
+	reader2, err := r.setUpReader()
 	if err != nil {
 		return errors.Trace(err)
 	}
+	defer func() {
+		err = reader2.Close()
+		if err != nil {
+			log.Errorf("[relay] close binlog event reader error %v", err)
+		}
+	}()
 
+	writer2, err := r.setUpWriter(parser2)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer func() {
+		err = writer2.Close()
+		if err != nil {
+			log.Errorf("[relay] close binlog event writer error %v", err)
+		}
+	}()
+
+	transformer2 := transformer.NewTransformer(parser2)
+
+	go r.doIntervalOps(parentCtx)
+
+	return errors.Trace(r.handleEvents(parentCtx, reader2, transformer2, writer2))
+}
+
+// tryRecoverLatestFile tries to recover latest relay log file with corrupt/incomplete binlog events/transactions.
+func (r *Relay) tryRecoverLatestFile(parser2 *parser.Parser) error {
 	var (
-		_, lastPos   = r.meta.Pos()
-		_, lastGTID  = r.meta.GTID()
-		tryReSync    = true  // used to handle master-slave switch
-		addRelayFlag = false // whether needing to add a LOG_EVENT_RELAY_LOG_F flag to the event
-
-		// fill gap steps:
-		// 1. record the pos after the gap
-		// 2. create a special streamer to fill the gap
-		// 3. catchup pos after the gap
-		// 4. close the special streamer
-		gapReader       binlogreader.Reader
-		gapSyncStartPos *mysql.Position                     // the pos of the event starting to fill the gap
-		gapSyncEndPos   *mysql.Position                     // the pos of the event after the gap
-		eventFormat     *replication.FormatDescriptionEvent // latest FormatDescriptionEvent, used when re-calculate checksum
+		uuid, latestPos = r.meta.Pos()
+		_, latestGTID   = r.meta.GTID()
 	)
 
-	closeGapSyncer := func() {
-		addRelayFlag = false // reset to false when closing the gap syncer
-		if gapReader != nil {
-			gapReader.Close()
-			gapReader = nil
-		}
-		gapSyncStartPos = nil
-		gapSyncEndPos = nil
+	if r.cfg.EnableGTID && latestPos.Compare(minCheckpoint) <= 0 {
+		log.Warnf("[relay] no corresponding position specified for GTID sets %s, skip recovering", latestGTID)
```

Comment:
what's meaning of it?

Reply:
if `enable-gtid` is `true`, user can only specify `relay-binlog-gtid` without `relay-binlog-name`, then we do not know which relay file should be recovered (the latest relay file is not accurate because the user can change `relay-binlog-gtid` to skip some relay log files).

Reply:
removed `r.cfg.EnableGTID` check in https://github.com/pingcap/dm/pull/171/commits/201f96028d56a0aaa1c17373381a391bdbe8e75a.